## 1. Загрузка данных и первичный анализ

### 1.1. Загрузка датасета

In [1]:
# Загрузите все необходимые библиотеки и датасет
# Выведите первые 5 строк датасета

import pandas as pd
import numpy as np

# Выводим первые 5 строк
df = pd.read_excel("3_Concrete_Data.xls")
print(df.head())

   Cement (component 1)(kg in a m^3 mixture)  \
0                                      540.0   
1                                      540.0   
2                                      332.5   
3                                      332.5   
4                                      198.6   

   Blast Furnace Slag (component 2)(kg in a m^3 mixture)  \
0                                                0.0       
1                                                0.0       
2                                              142.5       
3                                              142.5       
4                                              132.4       

   Fly Ash (component 3)(kg in a m^3 mixture)  \
0                                         0.0   
1                                         0.0   
2                                         0.0   
3                                         0.0   
4                                         0.0   

   Water  (component 4)(kg in a m^3 mixture)  \
0      

*(Кратко прокомментируйте, какие признаки есть в данных и что является целевой переменной.)*

### 1.2. Размерность и описательные статистики

In [14]:
# получите размерность датасета

print(f"Строк: {df.shape[0]}, Столбцов: {df.shape[1]}")

Строк: 1030, Столбцов: 9


In [15]:
# получите описательные статистики датасета

print(df.describe())

       Cement (component 1)(kg in a m^3 mixture)  \
count                                1030.000000   
mean                                  281.165631   
std                                   104.507142   
min                                   102.000000   
25%                                   192.375000   
50%                                   272.900000   
75%                                   350.000000   
max                                   540.000000   

       Blast Furnace Slag (component 2)(kg in a m^3 mixture)  \
count                                        1030.000000       
mean                                           73.895485       
std                                            86.279104       
min                                             0.000000       
25%                                             0.000000       
50%                                            22.000000       
75%                                           142.950000       
max                


Комментарий:

*Напишите 2–3 предложения по итогам первичного анализа.*

В датасете 1030 наблюдений и 9 столбцов. Пропуски отсутствуют, признаки имеют разный диапазон значений.

## 2. Признаки с низкой вариацией

### 2.1. Количество уникальных значений

In [16]:
# посчитайте количество уникальных значений
# выведите количество уникальных значений по столбцам

unique_counts = df.nunique()
print(unique_counts)

Cement (component 1)(kg in a m^3 mixture)                280
Blast Furnace Slag (component 2)(kg in a m^3 mixture)    187
Fly Ash (component 3)(kg in a m^3 mixture)               163
Water  (component 4)(kg in a m^3 mixture)                205
Superplasticizer (component 5)(kg in a m^3 mixture)      155
Coarse Aggregate  (component 6)(kg in a m^3 mixture)     284
Fine Aggregate (component 7)(kg in a m^3 mixture)        304
Age (day)                                                 14
Concrete compressive strength(MPa, megapascals)          938
dtype: int64


### 2.2. Поиск и удаление константных / почти константных признаков

In [17]:
# Найдите признаки с одной уникальной категорией
# Выведите их количество

counts = df.nunique()
to_del = [i for i, v in enumerate(counts) if v == 1]
print(to_del)

[]


In [19]:
# Приведите код для удаления колонок с константными признаками
# Выведите обновленную размерность датасета

df.drop(to_del, axis=1, inplace=True)
print(df.shape)

(1030, 9)


Комментарий:

*Объясните, почему эти признаки не полезны для модели.*

Признаки с одним уникальным значением не несут полезной информации, так как они одинаковы для всех объектов. Такие признаки не помогают модели различать объекты, но увеличивают размерность и замедляют обучение. Поэтому они удаляются на этапе предобработки данных.

## 3. Дубликаты строк

### 3.1. Поиск дубликатов


In [ ]:
# приведите код для поиска дубликатов строк
# выведите общее количество дубликатов строк в датасете

dup = df.duplicated(keep=False)
dups = df[dup].sort_values(list(df.columns))
print(dups)


     Cement (component 1)(kg in a m^3 mixture)  \
801                                      252.0   
809                                      252.0   
83                                       362.6   
86                                       362.6   
88                                       362.6   
91                                       362.6   
106                                      362.6   
109                                      362.6   
111                                      362.6   
129                                      362.6   
132                                      362.6   
134                                      362.6   
137                                      362.6   
152                                      362.6   
155                                      362.6   
157                                      362.6   
160                                      362.6   
175                                      362.6   
177                                      362.6   


In [4]:
# При наличии дубликатов выведите первые 5 строк из повторяющихся значений

dup = df.duplicated(keep=False)
dups = df[dup].sort_values(list(df.columns))
print(dups.head(5))

     Cement (component 1)(kg in a m^3 mixture)  \
801                                      252.0   
809                                      252.0   
83                                       362.6   
86                                       362.6   
88                                       362.6   

     Blast Furnace Slag (component 2)(kg in a m^3 mixture)  \
801                                                0.0       
809                                                0.0       
83                                               189.0       
86                                               189.0       
88                                               189.0       

     Fly Ash (component 3)(kg in a m^3 mixture)  \
801                                         0.0   
809                                         0.0   
83                                          0.0   
86                                          0.0   
88                                          0.0   

     Water  (compon

### 3.2. Удаление дубликатов

In [17]:
# Удалите дубликаты из датасета
# Проверьте размерность датасета после удаления дубликатов

Комментарий:

*Опишите, как дубликаты могут повлиять на обучение модели.*

## 4. Выбросы

### 4.1. Исследование распределений (по выбору 2–3 признаков)

In [18]:
# Приведеите пример гистограммы минимум 3х признаков

### 4.2. Определение выбросов (3σ или IQR)

In [29]:
# определите верхние и нижние пределы выбросов одним из способов для одного из признаков. 


In [30]:
# посчитайте количество выбросов


*(Повторите аналогичную процедуру для двух других признаков.)*

## 5. Пропуски в данных

### 5.1. Анализ пропусков

In [28]:
# посчитайте количество пропусков по столбцам


### 5.2. Стратегии обработки пропусков

#### Стратегия 1: (например, удаление строк с пропусками)

In [34]:
# приведите код для удаления пропусков
# выведите размер датасета послед удаления пропусков



#### Стратегия 2: (например, простая импутация средним / kNN / IterativeImputer)

In [39]:
# примените один из методов для замены пропущенных значений


Комментарий:

*Обоснуйте выбор стратегии и прокомментируйте ожидаемые эффекты.*

## 6. Построение и оценка моделей

### 6.1. Разбиение на обучающую и тестовую выборки

In [40]:
# импортируйте модуль из библиотеки sklearn для разбиения на обучающую и тестовую выборки
# разбейте все данные на тестовую и обучающую выборки


### 6.2. Модель 1: линейная регрессия

In [45]:
# обучите модель линейной регрессии и посчитайте MAE, RMSE, R2


Комментарий:

*Сделайте краткий вывод о качестве линейной модели.*

### 6.3. Модель 2: более сложная (например, RandomForestRegressor)

In [46]:
# обучите модель случайного леса и посчитайте MAE, RMSE, R2


Комментарий:

*Сравните результаты с линейной моделью. Какие преимущества/недостатки вы видите?*